src/metabase/driver/clickhouse.clj

```diff
@@ -112,6 +119,74 @@
            :semantic-version {:major (.getInt rset 2)
                               :minor (.getInt rset 3)}})))))
 
+(defmethod driver/upload-type->database-type :clickhouse
+  [_driver upload-type]
+  (case upload-type
+    ::upload/varchar-255              "Nullable(String)"
+    ::upload/text                     "Nullable(String)"
+    ::upload/int                      "Nullable(Int64)"
+    ::upload/float                    "Nullable(Float64)"
+    ::upload/boolean                  "Nullable(Boolean)"
+    ::upload/date                     "Nullable(Date32)"
+    ::upload/datetime                 "Nullable(DateTime64(3))"
+    ;; FIXME: should be `Nullable(DateTime64(3))`
+    ::upload/offset-datetime          nil))
+
+(defmethod driver/table-name-length-limit :clickhouse
+  [_driver]
+  ;; FIXME: This is a lie because you're really limited by a filesystems' limits, because Clickhouse uses
+  ;; filenames as table/column names. But its an approximation
+  206)
+
+(defn- quote-name [s]
+  (let [parts (str/split (name s) #"\.")]
+    (str/join "." (map #(str "`" % "`") parts))))
+
+(defn- create-table!-sql
+  [driver table-name column-definitions & {:keys [primary-key]}]
+  (str/join "\n"
+            [(first (sql/format {:create-table (keyword table-name)
+                                 :with-columns (mapv (fn [[name type-spec]]
+                                                       (vec (cons name [[:raw type-spec]])))
+                                                     column-definitions)}
+                                :quoted true
+                                :dialect (sql.qp/quote-style driver)))
+             "ENGINE = MergeTree"
+             (format "PRIMARY KEY (%s)" (str/join ", " (map quote-name primary-key)))
+             "ORDER BY ()"]))
+
+(defmethod driver/create-table! :clickhouse
+  [driver db-id table-name column-definitions & {:keys [primary-key]}]
+  (let [sql (create-table!-sql driver table-name column-definitions :primary-key primary-key)]
+    (qp.writeback/execute-write-sql! db-id sql)))
+
+(defmethod driver/insert-into! :clickhouse
```

Comment:
Rather than use the default sql-jdbc implementation, I've chosen to use the approach recommended [here](https://github.com/ClickHouse/clickhouse-java/blob/c25ffc62c886c0cdc1fcf8d8bf2234690c131dca/examples/jdbc/src/main/java/com/clickhouse/examples/jdbc/Basic.java#L74-L94), to allow for large uploads.